sormas-backend/src/main/java/de/symeda/sormas/backend/user/KeycloakService.java

```diff
@@ -0,0 +1,249 @@
+
+/**
+ * @author Alex Vidrean
+ * @since 15-Aug-20
+ */
+@Stateless
+@LocalBean
+public class KeycloakService {
+
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+
+	private static final String OIDC_REALM = "realm";
+	private static final String OIDC_SERVER_URL = "auth-server-url";
+	private static final String OIDC_CREDENTIALS = "credentials";
+	private static final String OIDC_SECRET = "secret";
+
+	private static final String REALM_NAME = "SORMAS";
+
+	private Keycloak keycloak = null;
+
+	@PostConstruct
+	public void init() {
+		Optional<String> oidcJson = ConfigProvider.getConfig().getOptionalValue("sormas.backend.security.oidc.json", String.class);
+
+		if (!oidcJson.isPresent()) {
+			logger.warn("Undefined KEYCLOAK configuration for sormas.backend.security.oidc.json. Configure the property or disable the KEYCLOAK authentication provider.");
+			return;
+		}
+
+		try {
+			JSONObject json = JSONObjectUtils.parse(oidcJson.get());
+
+			keycloak = KeycloakBuilder.builder()
+				.realm(json.getAsString(OIDC_REALM))
+				.serverUrl(json.getAsString(OIDC_SERVER_URL))
+				.clientId("sormas-backend")
+				.clientSecret(JSONObjectUtils.getJSONObject(json, OIDC_CREDENTIALS).getAsString(OIDC_SECRET))
+				.grantType(OAuth2Constants.CLIENT_CREDENTIALS)
+				.build();
+
+		} catch (ParseException e) {
+			throw new IllegalArgumentException("Invalid JSON for backend keycloak oidc");
+		}
+	}
+
+	public void handleUserCreateEvent(@Observes UserCreateEvent userCreateEvent) {
+		Optional<Keycloak> keycloak = getKeycloak();
+		if(!keycloak.isPresent()) {
+			logger.warn("Cannot obtain keycloak instance. Will not create user in keycloak");
+			return;
+		}
+
+		User user = userCreateEvent.getUser();
+		String userId = createUser(keycloak.get(), user);
+		sendActivationEmail(keycloak.get(), userId);
+	}
+
+	public void handleUserUpdateEvent(@Observes UserUpdateEvent userUpdateEvent) {
+		Optional<Keycloak> keycloak = getKeycloak();
+		if(!keycloak.isPresent()) {
+			logger.warn("Cannot obtain keycloak instance. Will not update user in keycloak");
+			return;
+		}
+
+		User newUser = userUpdateEvent.getNewUser();
+		User oldUser = userUpdateEvent.getOldUser();
+
+		Optional<UserRepresentation> userRepresentation = updateUser(keycloak.get(), oldUser, newUser);
+		if(!userRepresentation.isPresent()) {
+			logger.debug("Cannot find user in Keycloak. Will try to create it");
+			String userId = createUser(keycloak.get(), newUser);
+			sendActivationEmail(keycloak.get(), userId);
+		}
+	}
+
+	public void handlePasswordResetEvent(@Observes PasswordResetEvent passwordResetEvent) {
+		Optional<Keycloak> keycloak = getKeycloak();
+		if(!keycloak.isPresent()) {
+			logger.warn("Cannot obtain keycloak instance. Will not reset user password in keycloak");
+			return;
+		}
+
+		User user = passwordResetEvent.getUser();
+		Optional<UserRepresentation> userRepresentation = getUserByUsername(keycloak.get(), user.getUserName());
```

Comment:
for the update you are logging in case there is no user with this name - shouldn't it be logged here as well?
also this method to get the keycloak user, just for me to understand does keycloak allow duplicate user names for the same realm ? or is it a configuration ?

Reply:
Good point, I will add a log here as well.

It's a big strange because Keycloak doesn't allow duplicate usernames for the same realm, but it only offers a search endpoint to retrieve users by username which can be used to also search for substrings of the username, that's why the result is always returned as a list.

In our case however we are using `exact=true` which means that it searches for the exact username, but API still returns the response as a list.